include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -1567,6 +1567,207 @@ namespace internal
 
 
 
+  template <int dim,
+            int n_rows_static,
+            int n_columns_static,
+            typename Number,
+            typename Number2,
+            int  direction,
+            bool contract_over_rows,
+            bool add,
+            int  type,
+            bool one_line>
+  inline void
+  even_odd_apply(const int                       n_rows_in,
+                 const int                       n_columns_in,
+                 const Number2 *DEAL_II_RESTRICT shapes,
+                 const Number *                  in,
+                 Number *                        out)
+  {
+    static_assert(type < 3, "Only three variants type=0,1,2 implemented");
+    static_assert(one_line == false || direction == dim - 1,
+                  "Single-line evaluation only works for direction=dim-1.");
+
+    const int n_rows = n_rows_static == -1 ? n_rows_in : n_rows_static;
+    const int n_columns =
+      n_columns_static == -1 ? n_columns_in : n_columns_static;
+
+    Assert(dim == direction + 1 || one_line == true || n_rows == n_columns ||
+             in != out,
+           ExcMessage("In-place operation only supported for "
+                      "n_rows==n_columns or single-line interpolation"));
+
+    // We cannot statically assert that direction is less than dim, so must do
+    // an additional dynamic check
+    AssertIndexRange(direction, dim);
+
+    const int nn     = contract_over_rows ? n_columns : n_rows;
+    const int mm     = contract_over_rows ? n_rows : n_columns;
+    const int n_cols = nn / 2;
+    const int mid    = mm / 2;
+
+    const int stride    = Utilities::pow(n_columns, direction);
+    const int n_blocks1 = one_line ? 1 : stride;
+    const int n_blocks2 =
+      Utilities::pow(n_rows, (direction >= dim) ? 0 : (dim - direction - 1));
+
+    const int offset = (n_columns + 1) / 2;
+
+    // this code may look very inefficient at first sight due to the many
+    // different cases with if's at the innermost loop part, but all of the
+    // conditionals can be evaluated at compile time because they are
+    // templates, so the compiler should optimize everything away
+    for (int i2 = 0; i2 < n_blocks2; ++i2)
+      {
+        for (int i1 = 0; i1 < n_blocks1; ++i1)
+          {
+            Number xp[mid > 0 ? mid : 1], xm[mid > 0 ? mid : 1];
```

Comment:
Does this work? This would create an variable-sized array, which is (IIRC) a GNU extension.

Reply:
You are right. I am a bit surprised that I did not get a warning. I have pushed a fix.